presto-main/src/main/java/io/prestosql/operator/StreamingAggregationOperator.java

```diff
@@ -232,17 +230,25 @@ private StreamingAggregation(
                 if (outputPages.isEmpty()) {
                     return finished();
                 }
+
+                return ofResult(outputPages.removeFirst(), false);
             }
-            else {
-                processInput(inputPage);
-                updateMemoryUsage();
+
+            // flush remaining output pages before requesting next input page
+            if (!outputPages.isEmpty()) {
+                Page outputPage = outputPages.removeFirst();
+                return ofResult(outputPage, outputPages.isEmpty());
```

Comment:
i think this should be 
```suggestion
                return ofResult(outputPage, false);
```

-- current `inputPage` isn't consumed, so we should request it to be re-submitted

Reply:
> * When we exhausted output pages queue (`outputPages.isEmpty()`), then we can ask for next page.

we already got the next page (the `inputPage` argument). so it seems to me we can lose an input

Reply:
the `inputPage` stays until you ask for next page. You ask for next page only when there are no pending output pages.
Keep note that `outputPages` buffer needs to be empty in order to process next `inputPage`